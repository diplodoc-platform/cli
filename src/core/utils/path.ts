import {dirname, isAbsolute, join, normalize} from 'node:path';
import _normalizePath from 'normalize-path';

import {isExternalHref} from '~/core/utils/url';

export function normalizePath(path: string): NormalizedPath {
    return _normalizePath(normalize(_normalizePath(path, false)), false);
}

export function isRelativePath(path: string): path is RelativePath {
    return !isExternalHref(path) && !isAbsolute(path);
}

export function langFromPath(path: string, config: {lang?: string; langs: string[]}) {
    const {lang, langs} = config;
    const pathBaseLang = normalizePath(path).split('/')[0];
    const pathLang = langs.includes(pathBaseLang) && pathBaseLang;

    return pathLang || lang || langs[0];
}

export function rebasePath(root: RelativePath, path: RelativePath) {
    return normalizePath(join(dirname(root), path));
}

export function fullPath(
    path: AbsolutePath | NormalizedPath,
    root: NormalizedPath,
): NormalizedPath {
    if (path.match(/^(\/|\\)/)) {
        return normalizePath(path.slice(1));
    } else {
        return normalizePath(join(dirname(root), path));
    }
}
