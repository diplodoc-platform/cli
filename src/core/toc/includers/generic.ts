import type {Build} from '~/commands';
import type {IncluderOptions, RawToc, RawTocItem, YfmString} from '~/core/toc';

import {dirname, extname, join} from 'node:path';

// const AUTOTITLE = '{$T}';

type Options = IncluderOptions<{
    input?: RelativePath;
    autotitle?: boolean;
    leadingPage?: {
        autotitle?: boolean;
        name?: string;
    };
}>;

type Graph = {
    [prop: string]: (YfmString & NormalizedPath) | Graph;
};

// TODO: implement autotitle after md refactoring
// TODO: implement sort
export class GenericIncluderExtension {
    apply(program: Build) {
        program.hooks.BeforeAnyRun.tap('GenericIncluder', (run) => {
            run.toc.hooks.Includer.for('generic').tapPromise(
                'GenericIncluder',
                async (toc, options: Options, path) => {
                    const input = options.input
                        ? join(dirname(path), options.input)
                        : dirname(options.path);
                    const files = await run.glob('**/*.md', {
                        cwd: join(run.input, input),
                    });

                    return fillToc(toc, graph(files), options);
                },
            );
        });
    }
}

function graph(paths: NormalizedPath[]): Graph {
    const graph: Graph = {};

    for (const path of paths) {
        const chunks: string[] = path.split('/');

        let level: Hash = graph;
        while (chunks.length) {
            const field = chunks.shift() as string;

            if (chunks.length) {
                level[field] = level[field] || {};
                level = level[field];
            } else {
                level[field.replace(extname(field), '')] = path;
            }
        }
    }

    return graph;
}

function pageName(key: string, options: Options) {
    if (key === 'index') {
        // if (options?.leadingPage?.autotitle) {
        //     return AUTOTITLE;
        // }

        // TODO: i18n
        return (options?.leadingPage?.name ?? 'Overview') as YfmString;
    }

    // return options.autotitle ? AUTOTITLE : key;
    return key as YfmString;
}

function fillToc(toc: RawToc, graph: Graph, options: Options) {
    function item([key, value]: [string, Graph | (NormalizedPath & YfmString)]): RawTocItem {
        const name = pageName(key, options);

        if (typeof value === 'string') {
            return {name, href: value};
        }

        return {name, items: Object.entries(value).map(item)};
    }

    toc.items = Object.entries(graph).map(item);

    return toc;
}
