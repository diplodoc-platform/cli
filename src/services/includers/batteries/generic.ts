import {mkdir, readFile, writeFile} from 'fs/promises';
import {dirname, join, parse} from 'path';

import {updateWith} from 'lodash';
import {dump} from 'js-yaml';

import {glob} from '../../../utils/glob';

import {IncluderFunctionParams} from '../../../models';

class GenericIncluderError extends Error {
    path: string;

    constructor(message: string, path: string) {
        super(message);

        this.name = 'GenericIncluderError';
        this.path = path;
    }
}

const name = 'generic';

const MD_GLOB = '**/*.md';

type Params = {
    input: string;
    leadingPage: {
        name?: string;
    };
};

async function includerFunction(params: IncluderFunctionParams<Params>) {
    const {
        readBasePath,
        writeBasePath,
        tocPath,
        item,
        passedParams: {input, leadingPage},
        index,
    } = params;

    if (!input?.length || !item.include?.path) {
        throw new GenericIncluderError('provide includer with input parameter', tocPath);
    }

    try {
        const leadingPageName = leadingPage?.name ?? 'Overview';

        const tocDirPath = dirname(tocPath);

        const contentPath =
            index === 0
                ? join(writeBasePath, tocDirPath, input)
                : join(readBasePath, tocDirPath, input);

        let cache = {};
        let found = [];

        ({
            state: {found, cache},
        } = await glob(MD_GLOB, {
            cwd: contentPath,
            nosort: true,
            nocase: true,
            cache,
        }));

        const writePath = join(writeBasePath, tocDirPath, item.include.path);

        await mkdir(writePath, {recursive: true});

        for (const filePath of found) {
            const file = await readFile(join(contentPath, filePath));

            await mkdir(dirname(join(writePath, filePath)), {recursive: true});
            await writeFile(join(writePath, filePath), file);
        }

        const graph = createGraphFromPaths(found);

        const toc = createToc(leadingPageName, item.include.path)(graph, []);

        await writeFile(join(writePath, 'toc.yaml'), dump(toc));
    } catch (err) {
        throw new GenericIncluderError(err.toString(), tocPath);
    }
}

function createGraphFromPaths(paths: string[]) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const graph: Record<string, any> = {};

    for (const path of paths) {
        const chunks = path.split('/').filter(Boolean);
        if (chunks.length < 2) {
            if (chunks.length === 1) {
                graph.files = graph.files ? graph.files.concat(chunks[0]) : chunks;
            }

            continue;
        }

        const file = chunks.pop();

        updateWith(
            graph,
            chunks,
            (old) => {
                return old ? {files: [...old.files, file]} : {files: [file]};
            },
            Object,
        );
    }

    return graph;
}

function createToc(leadingPageName: string, tocName: string) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function createTocRec(
        graph: Record<string, any>,
        cursor: string[],
    ): Record<string, any> {
        const handler = (file: string) => ({
            name: parse(file).name === 'index' ? leadingPageName : file,
            href: join(...cursor, file),
        });

        const recurse = (key: string) => createTocRec(graph[key], [...cursor, key]);

        const current = {
            name: cursor[cursor.length - 1] ?? tocName,
            items: [
                ...(graph.files ?? []).map(handler),
                ...Object.keys(graph)
                    .filter((key) => key !== 'files')
                    .map(recurse),
            ],
        };

        return current;
    };
}

export {name, includerFunction};

export default {name, includerFunction};
