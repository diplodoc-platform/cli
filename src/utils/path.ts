import {sep} from 'path';
import {Platforms} from '../constants';

export function addSlashPrefix(path: string): string {
    const slashPrefix = path.startsWith(sep) ? '' : sep;

    return `${slashPrefix}${path}`;
}

export function convertBackSlashToSlash(path: string): string {
    if (process.platform === Platforms.WINDOWS) {
        return path.replace(/\\/g, '/');
    }

    return path;
}
