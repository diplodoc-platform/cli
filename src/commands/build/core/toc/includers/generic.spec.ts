import {describe, expect, it} from 'vitest';
import {when} from 'vitest-when';
import {join} from 'node:path';
import {dump} from 'js-yaml';

import {setupBuild, setupRun} from '~/commands/build/__tests__';

import {GenericIncluderExtension} from './generic';

const prepareExtension = async (globs: [string, RelativePath, NormalizedPath[]][]) => {
    const build = setupBuild();
    const run = setupRun({});
    const extension = new GenericIncluderExtension();

    for (const [pattern, cwd, files] of globs) {
        when(run.glob)
            .calledWith(
                pattern,
                expect.objectContaining({
                    cwd: join(run.input, cwd),
                }),
            )
            .thenResolve(files);
    }

    extension.apply(build);

    await build.hooks.BeforeAnyRun.promise(run);

    return {build, run, extension};
};

describe('Generic includer', () => {
    it('should work', async () => {
        const {run} = await prepareExtension([
            [
                '**/*.md',
                './test',
                [
                    'index.md',
                    'test.md',
                    'sub/sub-1.md',
                    'sub/sub-2.md',
                    'sub/sub/sub-3.md',
                    'skip/sub/sub-1.md',
                ] as NormalizedPath[],
            ],
        ]);

        const result = await run.toc.hooks.Includer.for('generic').promise(
            {},
            {
                input: './test',
                path: './test/toc.yaml',
            },
            './toc.yaml',
        );

        expect(dump(result)).toMatchSnapshot();
    });

    it('should use top path as input root, if input is not specified', async () => {
        const {run} = await prepareExtension([
            [
                '**/*.md',
                './path/test',
                [
                    'index.md',
                    'test.md',
                    'sub/sub-1.md',
                    'sub/sub-2.md',
                    'sub/sub/sub-3.md',
                    'skip/sub/sub-1.md',
                ] as NormalizedPath[],
            ],
        ]);

        const result = await run.toc.hooks.Includer.for('generic').promise(
            {},
            {
                path: './path/test/toc.yaml',
            },
            './toc.yaml',
        );

        expect(dump(result)).toMatchSnapshot();
    });

    it('should use leadingPage.name option', async () => {
        const {run} = await prepareExtension([
            [
                '**/*.md',
                './test',
                [
                    'index.md',
                    'test.md',
                    'sub/sub-1.md',
                    'sub/sub-2.md',
                    'sub/sub/sub-3.md',
                    'skip/sub/sub-1.md',
                ] as NormalizedPath[],
            ],
        ]);

        const result = await run.toc.hooks.Includer.for('generic').promise(
            {},
            {
                input: './test',
                path: './test/toc.yaml',
                leadingPage: {
                    name: 'Test Overview',
                },
            },
            './toc.yaml',
        );

        expect(dump(result)).toMatchSnapshot();
    });
});
