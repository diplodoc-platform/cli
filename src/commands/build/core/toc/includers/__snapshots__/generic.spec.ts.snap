// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generic includer > should use leadingPage.name option 1`] = `
"items:
  - name: Test Overview
    href: index.md
  - name: test
    href: test.md
  - name: sub
    items:
      - name: sub-1
        href: sub/sub-1.md
      - name: sub-2
        href: sub/sub-2.md
      - name: sub
        items:
          - name: sub-3
            href: sub/sub/sub-3.md
  - name: skip
    items:
      - name: sub
        items:
          - name: sub-1
            href: skip/sub/sub-1.md
"
`;

exports[`Generic includer > should use top path as input root, if input is not specified 1`] = `
"items:
  - name: Overview
    href: index.md
  - name: test
    href: test.md
  - name: sub
    items:
      - name: sub-1
        href: sub/sub-1.md
      - name: sub-2
        href: sub/sub-2.md
      - name: sub
        items:
          - name: sub-3
            href: sub/sub/sub-3.md
  - name: skip
    items:
      - name: sub
        items:
          - name: sub-1
            href: skip/sub/sub-1.md
"
`;

exports[`Generic includer > should work 1`] = `
"items:
  - name: Overview
    href: index.md
  - name: test
    href: test.md
  - name: sub
    items:
      - name: sub-1
        href: sub/sub-1.md
      - name: sub-2
        href: sub/sub-2.md
      - name: sub
        items:
          - name: sub-3
            href: sub/sub/sub-3.md
  - name: skip
    items:
      - name: sub
        items:
          - name: sub-1
            href: skip/sub/sub-1.md
"
`;
